### ISSUE TITLE
[cli] Bug fix proposal:  Correct envTargetChoices filtering when existing targets include arrays

### ISSUE THREAD

**Issue by developerjhp** [zeit/now#12092]
### Background
Hi, I am preparing a PR to fix a bug in the `envTargetChoices` filtering logic. Currently, when `existingTargets` includes arrays, the resulting `choices` array becomes empty.

In the `src/commands/env/add.ts` file, there is a bug in the `envTargetChoices` filtering logic. When the `existingTargets` variable includes arrays (e.g., `['preview']` or `['production', 'preview', 'development']`), the `choices` array ends up empty because the logic flattens the `existingTargets` values into individual elements, causing all available choices to be filtered out. 

This issue was caused by changes made in [this commit](https://github.com/vercel/vercel/commit/ddcb7a1891b3b9dcaf5ffb1e15c31f0115f30db2). 

The issue is not with the getEnvRecords function itself, but with the logic below where existingTargets and existingCustomEnvs are being processed. Specifically, when iterating over matchingEnvs, if env.target is an array, each value in the array is being added individually to the Set. This causes the CLI to incorrectly filter out valid choices when adding environment variables.

The core problem arises because the values inside `existingTargets` are flattened into individual elements, rather than handling arrays as whole units. This results in the incorrect filtering of choices, leaving no valid options available.

### Proposed Change
- Modify the logic so that arrays are added to `existingTargets` as whole elements rather than flattening them into individual values.
- Update the filtering logic to correctly handle and compare values within arrays inside `existingTargets`.

```diff
- const existingTargets = new Set<string>();
- const existingCustomEnvs = new Set<string>();
- for (const env of matchingEnvs) {
-   if (typeof env.target === 'string') {
-     existingTargets.add(env.target);
-   } else if (Array.isArray(env.target)) {
-     for (const target of env.target) {
-       existingTargets.add(target);
-     }
-   }
-   if (env.customEnvironmentIds) {
-     for (const customEnvId of env.customEnvironmentIds) {
-       existingCustomEnvs.add(customEnvId);
-     }
-   }
- }

+ const existingTargets = new Set(matchingEnvs.map(env => env.target));
+ const existingCustomEnvs = new Set(
+   matchingEnvs.flatMap(env => env.customEnvironmentIds || [])
+ );

I believe this change resolves the issue and would appreciate your feedback on the proposed solution.

Thank you!

**Comment by TooTallNate** [comment]
Can you describe which scenario you are attempting to fix here? I'm not quite understanding.

> if env.target is an array, each value in the array is being added individually to the Set

This part sounds like the correct logic to me. If, say, "preview" is already defined as an env var, it should not be a selectable choice because there already exists this env var for preview and there can not be duplicates. Please let me know if I am missing something.

**Comment by trek** [comment]
@developerjhp could post steps to reproduce what you're seeing and the version of CLI you're using?

**Comment by developerjhp** [comment]
@TooTallNate 

Thank you for your response. Let me provide more details to help clarify the issue I’m seeing.

The logic of preventing duplicates when an environment variable is already defined for “preview” or other environments like “production” or “development” makes sense, and I agree that the values should not be selectable again.

However, the issue I’m experiencing occurs when I attempt to add the API_SERVER_URL environment variable to a specific preview branch (e.g., test-branch). The variable already exists in the production, preview, and development environments, and I want to add it specifically for test-branch in the preview environment.

Instead of allowing me to add the variable to this branch under preview, I receive the following error:
```
Error: The variable "API_SERVER_URL" has already been added to all Environments. To remove, run `vercel env rm API_SERVER_URL`
```

Upon investigating, I traced the issue to this commit: [ddcb7a1](https://github.com/vercel/vercel/commit/ddcb7a1891b3b9dcaf5ffb1e15c31f0115f30db2), where the existingTargets data structure was changed from Set<string[]> to Set<string>.

This change seems to cause an issue when trying to add environment variables to specific branches within the same environment. For example, when I try to add the API_SERVER_URL variable to test-branch (which is part of the preview environment), the logic:
```
envTargetChoices.filter(c => !existingTargets.has(c.value))
```
ends up returning an empty array because existingTargets contains preview, which already has the variable defined for other branches.

This causes the choices array to be empty, and then the following block gets triggered:
```js
if (!envGitBranch && choices.length === 0 && !opts['--force']) {
  output.error(
    `The variable ${param(envName)} has already been added to all Environments. To remove, run ${getCommandName(`env rm ${envName}`)}.`
  );
  return 1;
}
```

It seems like the change to the existingTargets data structure has impacted how the system handles branch-specific environment variables within the same environment (preview in this case).

**Comment by developerjhp** [comment]
@trek 

Sorry for the oversight. I'd like to add some crucial information I missed earlier

CLI VERSION : 37.4.0

For the past three months, the CLI has been working perfectly fine. However, starting about two weeks ago, I started seeing this error:
```
Error: The variable "API_SERVER_URL" has already been added to all Environments. To remove, run `vercel env rm API_SERVER_URL`.
```

This happens when I try to add the API_SERVER_URL environment variable to a test-branch in the preview environment. This is despite the fact that the API_SERVER_URL variable is already stored in production, preview, and development environments.

After investigating, I discovered that the issue seems to have started after the following commit: [ddcb7a1](https://github.com/vercel/vercel/commit/ddcb7a1891b3b9dcaf5ffb1e15c31f0115f30db2), specifically in the packages/cli/src/commands/env/add.ts file.

Previously, the existingTargets variable used a data structure of type Set<string[]>. However, in the new version, it was changed to Set<string>. As a result, when I attempt to add an environment variable that already exists in preview, development, and production to a test-branch under the preview environment, the following logic causes an issue:

The code envTargetChoices.filter(c => !existingTargets.has(c.value)) ends up returning an empty array, which causes the choices array to also be empty. Consequently, the following block of code gets triggered:
```
if (!envGitBranch && choices.length === 0 && !opts['--force']) {
  output.error(
    `The variable ${param(envName)} has already been added to all Environments. To remove, run ${getCommandName(`env rm ${envName}`)}.`
  );
  return 1;
}
```

This prevents me from adding the environment variable to the test-branch even though it should be allowed.
